client/python/openlineage/client/transport/http.py

```diff
@@ -70,10 +83,544 @@ def create_token_provider(auth: dict[str, str]) -> TokenProvider:
+    ):
+        with self.lock:
+            if run_id not in self.pending_completion_events:
+                self.pending_completion_events[run_id] = []
+            self.pending_completion_events[run_id].append((event_id, body, headers))
+
+    def is_start_in_progress(self, run_id: str) -> bool:
+        with self.lock:
+            start_event_id = f"{run_id}-START"
+            return start_event_id in self.events
+
+    def all_processed(self) -> bool:
+        with self.lock:
+            return len(self.events) == 0
+
+    def get_stats(self) -> dict[str, int]:
+        return self.stats
+
+
+class Emitter:
+    def emit(self, _: Event) -> None:
+        msg = "Subclasses must implement this method"
+        raise NotImplementedError(msg)
+
+    def wait_for_completion(self, _: float = 30.0) -> bool:
+        return True
+
+    def get_stats(self) -> dict[str, int]:
+        return {}
+
+    def shutdown(self, _: bool = True, timeout: float = 30.0) -> bool:
+        return self.wait_for_completion(timeout)
+
+    def _prepare_request(self, event: str) -> tuple[bytes | str, dict[str, str]]:
+        headers = {
+            "Content-Type": "application/json",
+            **self._auth_headers(self.config.auth),
+            **self.config.custom_headers,
+        }
+        if self.config.compression == HttpCompression.GZIP:
+            headers["Content-Encoding"] = "gzip"
+            return gzip.compress(event.encode("utf-8")), headers
+
+        return event, headers
+
+    def _auth_headers(self, token_provider: TokenProvider) -> dict:  # type: ignore[type-arg]
+        bearer = token_provider.get_bearer()
+        if bearer:
+            return {"Authorization": bearer}
+        return {}
+
+
+class AsyncEmitter(Emitter):
+    def __init__(self, config: HttpConfig) -> None:
+        self.config = config
+        # Create a queue with 2x the configured size to reserve space for released completion events
+        self.configured_queue_size = config.async_config.max_queue_size
+        self.event_queue = queue.Queue(maxsize=self.configured_queue_size * 2)
+        self.event_tracker = EventTracker()
+        self.should_exit = threading.Event()
+        # Maximum number of concurrent requests
+        self.max_concurrent = config.async_config.max_concurrent_requests
+        # Create a stopping event that will be set when we're shutting down
+        self.stopping = False
+        self.worker_thread = threading.Thread(target=self._setup_loop, daemon=True)
+        self.worker_thread.start()
+        log.debug(
+            "AsyncEmitter for %s started with configured queue size: %s (actual: %s), max concurrent: %s",
+            self.config.url,
+            self.configured_queue_size,
+            self.configured_queue_size * 2,
+            self.max_concurrent,
+        )
+
+    def _setup_loop(self) -> None:
+        self.loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self.loop)
```

Comment:
This makes event loop a global object, which may interfere with other async libraries of frameworks. Could we call methods like `loop.create_task(...)` directly?

Reply:
It's run in a worker thread, that only handles that loop. It is created by `AsyncEmitter`:

```
        self.worker_thread = threading.Thread(target=self._setup_loop, daemon=True)
```

Nothing else should run in that thread, or it should not reuse event loop from some other library. 

We can manage event loop manually but I don't know what would be the advantage of that in this case

Reply:
https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.set_event_loop
```
Set loop as the current event loop for the current OS thread.
```

Reply:
https://github.com/python/cpython/blob/main/Lib/asyncio/events.py#L706-L727

Reply:
Great, thanks! If you don't mind, please add a comment that this is a thread-local loop